static/src/javascripts/lib/cmp.spec.js

```diff
@@ -0,0 +1,215 @@
+// @flow
+import { init, onConsentNotification, consentState, _ } from 'lib/cmp';
+import { getAdConsentState as _getAdConsentState } from 'common/modules/commercial/ad-prefs.lib';
+
+const getAdConsentState: any = _getAdConsentState;
+
+jest.mock('common/modules/commercial/ad-prefs.lib', () => ({
+    getAdConsentState: jest.fn(),
+    thirdPartyTrackingAdConsent: jest.fn(),
+}));
+
+describe('cmp', () => {
+    beforeEach(() => {
+        _.resetCmp();
+        getAdConsentState.mockReset();
+    });
+
+    it('does not re-trigger consent notifications if init called multiple times', () => {
+        const myCallBack = jest.fn();
+
+        init();
+
+        onConsentNotification('functional', myCallBack);
+
+        expect(myCallBack).toHaveBeenCalledTimes(1);
+
+        init();
+
+        expect(myCallBack).toHaveBeenCalledTimes(1);
+    });
+
+    describe('consentState', () => {
+        it('returns null state if cmp lib has not been initialised', () => {
+            expect(consentState('functional')).toBeNull();
```

Comment:
This makes me wonder if the `consentState` functionality should even be exposed. Right now it's possible for a future developer that doesn't understand this code to just read the consent state from their library directly, without using the callbacks. In this case they'd likely just get `null` and might not know they need to use the listener approach to be told when the CMP library has loaded up.

What's the use case for exposing `consentState`?

Reply:
`consentState` is needed in such cases where asynchronous module isn't possible or parameter change isn't possible mid execution without major rework of the module's code. Such an example is the youtube player, whether the personalised ads flag is set when the player is started. Changes to the player can be made such that (for example) it reinitialises completely with the new state if a video isn't playing but that would mean a rework of that whole module, which is something we can't afford at the moment.  For this and other such cases we're sticking with checking the consent state at module init (same as it was before) and that's why we expose consentStat. Don't forget cmp lib's init is synchronous so everything after that will have the correct consent state already set in the lib.

Having said that this test's description is misleading. Perhaps it should be changed to something like "returns null if consent hasn't been set by the user" and the CMP lib README should clearly state that it should be run as early as possible to avoid module's calling it before it's ready.

Reply:
> this test's description is misleading. 

I disagree, it's possible to call `consentState` before the cmp lib has been initialised, and whilst this is possible we should understand how it'll behave, and this test demonstrates that. If you update this test to test "returns null if consent hasn't been set by the user" as suggested then it'd fail as right now it returns `true` for `functional` and `performance`.

Reply:
It's only misleading because it's not clear that even though we account for it calling `consentState()` before running `init()` is not the right way to use the library. Sorry for not being clear before.

Reply:
**Update**: Had a chat with @GHaberis and we decided to remove `init()` altogether to avoid confusion and make the library more robust. We'll make changes so the init logic is executed when a function from the library is first called.

Reply:
> We'll make changes so the init logic is executed when a function from the library is first called.
Much more sensible, nice one.

Reply:
> `consentState` is needed in such cases where asynchronous module isn't possible or parameter change isn't possible mid execution without major rework of the module's code. Such an example is the youtube player, whether the personalised ads flag is set when the player is started. 

We could  update the youtube initialisation step here https://github.com/guardian/frontend/blob/master/static/src/javascripts/bootstraps/enhanced/youtube.js#L6 so it's wrapped in a `onConsentNotification` call, and we pass the advertising consent state down like so:

```js
export const init = () => {
  onConsentNotification('advertisement', state => {
    checkElemsForVideos(state);
  });
};
```

Of course you'd only want to run `checkElemsForVideos` once. In which case you'd have to  add a flag of some sorts, as so:

```js
export const init = () => {
  let hasRun = false;
  onConsentNotification('advertisement', state => {
    if (!hasRun) {
      checkElemsForVideos(state);
      hasRun = true;
    }
  });
};
```

I also wonder whether  we want to build the idea of  only running once into the parameters of onConsentNotification,  for example add a third optional parameter  "runOnce"  (boolean, default = false). So you'd call it like:

```js
export const init = () => {
  onConsentNotification('advertisement', state => {
    checkElemsForVideos(state);
  }, true);
};
```

Reply:
I agree with your first suggestions to pass `checkElemsForVideos()` into a callback to `onConsentNotification()` but adding a runOnce boolean seems to go against what we want to do with consent, which is that each module should adapt every time consent is changed, not just the first time it changes.

Reply:
Apart from youtube only 2 other modules use `consentState()`. Those are `consent-tracker.js` and `user-ad-targeting.js`. The first we don't really need to worry about because it'll deleted when we start using the new consent audit log. The second can be converted to use a similar solution to the youtube one.

Which means that if no one can think of any other use cases we're safe to remove `consentState()`.
What do you think @GHaberis ?

Reply:
> I agree with your first suggestions to pass `checkElemsForVideos()` into a callback to `onConsentNotification()` but adding a runOnce boolean seems to go against what we want to do with consent, which is that each module should adapt every time consent is changed, not just the first time it changes.

The youtube initialisation is an example where we wouldn't want to adapt and run `checkElemsForVideos` on subsequent `onConsentNotification` events, these videos initialise once based on the first state given to them and that's it. In which case we have 2 options, the flag as suggested or some kind of `runOnce` parameter.

Reply:
> Apart from youtube only 2 other modules use `consentState()`. Those are `consent-tracker.js` and `user-ad-targeting.js`. The first we don't really need to worry about because it'll deleted when we start using the new consent audit log. The second can be converted to use a similar solution to the youtube one.
> 
> Which means that if no one can think of any other use cases we're safe to remove `consentState()`.
> What do you think @GHaberis ?

If we can remove for now then that sounds good to me 👍